lib/ex_doc/autolink.ex

```diff
@@ -0,0 +1,397 @@
+  defstruct [
+    :app,
+    :current_module,
+    :module_id,
+    :id,
+    ext: ".html",
+    skip_undefined_reference_warnings_on: []
+  ]
+
+  alias ExDoc.Formatter.HTML.Templates, as: T
+  alias ExDoc.Refs
+
+  @hexdocs "https://hexdocs.pm/"
+  @otpdocs "http://www.erlang.org/doc/man/"
+
+  def doc(ast, options \\ []) do
+    config = struct!(__MODULE__, options)
+    walk(ast, config)
+  end
+
+  defp walk(list, config) when is_list(list) do
+    Enum.map(list, &walk(&1, config))
+  end
+
+  defp walk(binary, _) when is_binary(binary) do
+    binary
+  end
+
+  defp walk({:pre, _, _} = ast, _config) do
+    ast
+  end
+
+  defp walk({:a, attrs, _} = ast, config) do
+    if url = custom_link(attrs, config) do
+      {:a, Keyword.put(attrs, :href, url), ast}
+    else
+      ast
+    end
+  end
+
+  defp walk({:code, _, [text]} = ast, config) do
+    case text_to_ref(text, :regular, config) do
+      :no_ref ->
+        ast
+
+      ref ->
+        if url = url(ref, config) do
+          {:a, [href: url], [ast]}
+        else
+          ast
+        end
+    end
+  end
+
+  defp walk({tag, attrs, ast}, config) do
+    {tag, attrs, walk(ast, config)}
+  end
+
+  defp custom_link(attrs, config) do
+    with {:ok, href} <- Keyword.fetch(attrs, :href),
+         [[_, text]] <- Regex.scan(~r/^`(.+)`$/, href) do
+      case text_to_ref(text, :custom_link, config) do
+        :no_ref ->
+          nil
+
+        ref ->
+          url(ref, config)
+      end
+    else
+      _ -> nil
+    end
+  end
+
+  defp text_to_ref("c:" <> text, mode, config) do
+    with {call, {:function, module, name, arity}} <- text_to_ref(text, mode, config) do
+      {call, {:callback, module, name, arity}}
+    end
+  end
+
+  defp text_to_ref("t:" <> text, mode, config) do
+    with {call, {:function, module, name, arity}} <- text_to_ref(text, mode, config) do
+      {call, {:type, module, name, arity}}
+    end
+  end
+
+  defp text_to_ref("mix help " <> name, _mode, _config), do: mix_task(name)
+  defp text_to_ref("mix " <> name, _mode, _config), do: mix_task(name)
+
+  defp text_to_ref(text, mode, config) do
+    if not String.contains?(text, [" ", "(", ")"]) do
+      case Code.string_to_quoted(text) do
+        {:ok, {:__aliases__, _, _} = module} ->
+          {:module, module(module)}
+
+        {:ok, {:/, _, [{{:., _, [module, name]}, _, []}, arity]}}
+        when is_atom(name) and is_integer(arity) ->
+          {:remote, {:function, module(module), name, arity}}
+
+        {:ok, {:/, _, [{name, _, _}, arity]}}
+        when is_atom(name) and is_integer(arity) ->
+          {:local, {:function, config.current_module, name, arity}}
+
+        {:ok, erlang_module} when is_atom(erlang_module) and mode == :custom_link ->
+          {:module, erlang_module}
+
+        _ ->
+          :no_ref
+      end
+    else
+      :no_ref
+    end
+  end
+
+  defp module({:__aliases__, _, _} = module), do: Module.concat([Macro.to_string(module)])
```

Comment:
I think this can be:

```suggestion
  defp module({:__aliases__, _, parts}), do: Module.concat(parts)
```